File: processor/src/test/resources/scenarios/mongocodecs/NonAccessibleMongoCodec.java
package scenarios.mongocodecs;

import org.bson.BsonReader;
import org.bson.BsonWriter;
import org.bson.codecs.Codec;
import org.bson.codecs.DecoderContext;
import org.bson.codecs.EncoderContext;

/**
 * Holds a private nested {@link Codec} that cannot be reached from any package.
 * Used to verify the visibility check rejects unconditionally inaccessible classes.
 */
public final class NonAccessibleMongoCodec {

    private NonAccessibleMongoCodec() {
    }

    private static final class Inner implements Codec<String> {

        public Inner() {
        }

        @Override
        public void encode(BsonWriter writer, String value, EncoderContext encoderContext) {
            writer.writeString(value);
        }

        @Override
        public String decode(BsonReader reader, DecoderContext decoderContext) {
            return reader.readString();
        }

        @Override
        public Class<String> getEncoderClass() {
            return String.class;
        }
    }
}

File: ksp/src/test/resources/scenarios/mongocodecs/NonAccessibleMongoCodec.kt
package scenarios.mongocodecs

import org.bson.BsonReader
import org.bson.BsonWriter
import org.bson.codecs.Codec
import org.bson.codecs.DecoderContext
import org.bson.codecs.EncoderContext

/**
 * Top-level `private` (file-scoped) class. Cannot be referenced from any other file,
 * regardless of module or package. Used to verify that the visibility check rejects
 * unconditionally inaccessible classes.
 */
private class NonAccessibleMongoCodec : Codec<String> {
    override fun encode(writer: BsonWriter, value: String, encoderContext: EncoderContext) {
        writer.writeString(value)
    }

    override fun decode(reader: BsonReader, decoderContext: DecoderContext): String =
        reader.readString()

    override fun getEncoderClass(): Class<String> = String::class.java
}
